draft session - 0 intro to programming (#4)
worldbank/dec-python-course: 0-intro-to-programming/slides-code/0-intro-to-programming.Rmd

  
  1. Some common concepts in programming 
]

---
class: middle

## What we assume about you:
.pull-left[
  
  Python can be used for countless of topics in computer science. 
  However, the scope of this training is limited to:
  
  1. How to use Python for data work. Numbers, text data, geo-data etc.
  
  1. Participants have some experience working with data. 
  For example, extensive work in Excel or some work in Stata or R.
  
  1. This course does not include how to use Python 
  for software development, backends for websites etc.
]

.pull-right[
  
  IMAGE OF EXCLE AND STATA LOGO WITH ARROW TOWARDS PYTHON LOGO
]

---
class: inverse, center, middle

# Why programming? 
# Work flow in data work programming

---
class:  middle

## Why use programming for data work? 

.pull-left[

  Data work should be very exact. Computers are more exact than humans.

  Instead of manually working directly one the data, 
  we write code that are instructions for computers 
  on what to do with the data.
  
  Since computers are very exact, 
  our code becomes an exact record of what was done to the data 
  and how outputs were generated.
]

.pull-right[

]

---
class:  middle

> ## Compare Excel and programming

Author: I might have to use `css` to animate text in table. I will look in to this but not put it at the top of my list

> ## Compare Excel and programming

Author: I struggled how to make that happen, but instead i did the table less verbose in 13c213b66ae81ee34e6ca75fe9b4be1e52186c3d and will go with that compromise

  | | Excel | Programming |
  |---------|-----|-----|
  | Are manual edits done directly to the files containing data? | Yes, we manually work directly in the file containing the raw data and and outputs | We only manually edit the code which is in files from files with data and output.  |
  | Can we write instructions for the computer to modify our data? | Yes, this is an option (Excel formulas) | Yes, and this is the only way (code)  |

>   | Can the instructions to modify the data be mixed with manual edits? | Yes, and it is common. | No, code should always start from unaltered raw data |

Author: ```suggestion
  | Can computer instructions be mixed with manual edits? | Yes, and it is common. | No, code should always start from unaltered raw data |
```

  | Do we have an unambiguous way of sharing what was done to the data? | Only formulas are shown, not manual edits | Yes, it is the code |
  | Is it easy for a reviewer to reproduce all steps? | Its challenging or impossible to follow any manual edits as there is no record of them | Yes, anyone who has the code and have access to the input data can reproduce everything |

---
class: middle

## Programming is like writing recepies

.pull-left[
  ### Sponge cake:
  * Set oven on 200C
  * Melt 200g butter and mix with 200g sugar
  * Mix in 4 eggs - one at the time
  * Mix in 200g flour
  * Bake in the middle of the oven for 30 min
]

.pull-right[
  ### Data science visualization:
  * Get GDP per capita data from for all countries
  * Get investment data from all WB client countries
  * Remove data that was not found in both dataset
  * Sort ascendingly on GDP per capita
  * Export a scatter plot and format the colors and the text
]

---
class: middle

## Data work in programmaing

.pull-left[

Ingredients in a cooking recipes can vary.

> Inputs to code can and should be very exact. 

Author: Ok, thanks. I added that here in the slides: 50a470468303dae5d78a618331e69b46fbc55ae5

  
  Data scientists use code to communicate about data work.
  
  A cook book of poorly written recipes will lead to bad food. 
  
  Poorly written code will lead to bad communication about data work
]

.pull-right[PICTURE]

---
class: middle, center

.large[
  .margin-bottom100[
    Code is not just a means to an end - **code is an end in itself**
  ]
  
  The code is a comprehensive and exact method
  to communicate the project's data work, 
  and **the code is an output** just as much as 
  the project's final report, data visualizations etc.
]

---
class: inverse, center, middle

# Why Python?

---

class: 

## Why Python?

.pull-left[

Why are there so many programming languages? How does Python compare to other languages?

**Free/Open-source**: Python is and will always be 100% free to use

**Easy syntax**: Python code reads more similarly to regular English

**Widely used**: Many users means many user developed tools

**#1 in data science**: The language with most data science tools

**General purpose**: Python can be used for data science, servers, software etc.
]

.pull-right[

]

---
class: inverse, center, middle

# How does one write good code?

---

> 

Author: It was actually something redundant - removed in: 752df50686bbd9c56fc150c3e2cabaeede9954ff

---
class: middle

## Learn how to commnuicate data science well

.pull-left[
  **Correct code:** Code that generates correct results
  
  **Good code:** Code that can be understood and used by others
  
  Code should be **both** correct code and good code.
  Only then is code useful as an output of your data work.
  
  Start by writing _correct code_, 
  then keep going to write _good code_.
]

.pull-right[
]

---
class: middle

## Trial and error!

**Writing correct code usually takes many errors**

> Don't be afraid of making errors as the data is not edited the data directly! 

Author: Yes, and there were missing words in the sentence after. Sorry. Fixed in: 752df50686bbd9c56fc150c3e2cabaeede9954ff

Even the best coders user trial and error - what makes them good is how 

---

class: middle

## Errors and bugs

.pull-left[
* Errors is a part of programming and they should be your friend
  
* Python always give you some message when there is an error
  
* Learn how to interpret error messages - Google them

* You will have thousands of errors and bugs - 
test run your code and find out where it breaks
]

.pull-right[
  PICTURE TEXT ERROR IN HEART
]

---
class: middle

## When the code works you are only half-done

How to make correct code into good code?

Ask yourself 
"_Would someone with your level of Python knowledge
be able to follow the code?_"

* Simplify
* Peer-review
* Document

Settling for "_just_" correct code does not 
save you or your team any time in the long run.

---

class: middle

## Re-use code

Since the instructions (code) is separate from input and output,
we can re-use code blocks within and across projects. An imporant paradigm in coding is **DRY** (don't uepeat yourself)

Three levels of re-using code:

> * **Copy-paste:** Copy a block a code of modify it to fit the context

Author: ```suggestion
* **Copy-paste:** Copy someone else's code and modify it to fit your context
```
I meant to talk about best-practice here, but I think your comment first helped me realize it was ambiguous and I am suggesting this edit. 

This comment was not meant to suggest copy-paste within your own code. For that a function should always be preferred. This was more a "do not re-invent the wheel" slide to show the power of building on other people's work. Maybe I need a DRY slide after this to make both points.

> * **Copy-paste:** Copy a block a code of modify it to fit the context

Author: Let me know if you think that is a good solution

> * **Copy-paste:** Copy a block a code of modify it to fit the context

Reviewer: I see. Yes, I had understood the point in the other sense of copying and pasting your own code. I think this is clear now with your edit

> * **Copy-paste:** Copy a block a code of modify it to fit the context

Author: Implemented in 3c1807407d2816f3a0bcc6033ded8fd78119f286 instead of here

]

---
class: middle

## Learn good code from peer coders
.pull-left[
  Read code from real world data work examples, 
  written by people at or just above your skill level
  
  This will teach you lessons 
  much more applicable to your day-to-day work,
  than school book examples
  
  Code written by someone at your level 
  will teach you much more about how you
  can become better at writing good code.
  
]

---
class: middle,

## Documentation - comments

.pull-left[

  All programming languages has a feature called _comments_. 
  
  Comments are text in the same file as the code 
  that the computer skips when running the code.
  
  Comments' only (but important) purpose
  is to provide context for a human reading the code.
  
  Comments best compliment the code if 
  they explains why the code is written a certain way.
]

.pull-right[

  | OK comment | Good comment |
  | --- | --- |
  | _What_ code does | _What_ code does **and** _why_ |
  | `Load data` | `Loading region definitions shared by NGO used in sampling` |
  | `Deleting variable "id" and "name"` | `Removing identifying variables to create a non-confidential data set` |
  
]

---
class: inverse, center, middle

> # Some common concepts in programming 

Reviewer: for this section I suggest adding a slide explaining what a script is and moving version control after code flow

---
class: middle

> ## Version control

Reviewer: I don't know if that space left to the right side of the slide is for images you're pending to add, but perhaps you can add a screenshot of a series of commits or changes displayed on GH to help explain this concept

Author: I had just the logo in mind, but I like your suggestion better.

We use **if-conditions** to instruct the computer
when to skip some code. 

* Only execute certain blocks on certain countries in a loop

* Code intended to be re-used across projects can fit more cases
when **if-conditions** are used
to make the code adapt itself to what is needed in each context.

]

.pull-right[
  ```{r, eval=TRUE, echo=FALSE,out.width = '75%'}
    knitr::include_graphics("img/code-flow-if.png")
  ```
]

---
class: middle

.pull-left[

  ## Code flow - combine it all
  
  Often loops and if-conditions are combined or they can be nested. 
  
  Session 2 of this course will show how this is done in Python. 
  
  Having neat, human readable and intuitive code flow 
  using loops and if-conditions,
  is important to make code _good code_.
]

.pull-right[
  ```{r, eval=TRUE, echo=FALSE,out.width = '75%'}
    knitr::include_graphics("img/code-flow-example.png")
  ```
] 

---
class: middle

## Summary in terms of research standards

**1. Transparency**<br>
You can share your code as a fully transparent log
of everything ever done to the input data.
This log can be shared with others,
but the person likely to benefit from this most frequently is yourself.

> **2. Replication**<br>

Author: Thanks for catching this. I left titles unedited after editing the content. It is updated in 7d5c1d60c87a12910c711b322145c51e40eca610

      knitr::include_graphics("img/ihearterrors.png")
    ```
  ]
] 

.pull-right[

## Errors and bugs

* Errors is a part of programming and they should be your friend
  
* Python always give you some message when there is an error
  
* Learn how to interpret error messages - Google them

* You will have thousands of errors and bugs - 
test run your code and find out where it breaks
]

---
class: middle

.pull-left[
  .center[
   # When your code works<br>you are only half-done
  ]
  .right[
    Ancient proverb
  ]
]

.pull-right[

  ### Path towards good code
  
  How to make correct code into good code?
  
  Ask yourself 
  "_Would someone with your level of Python knowledge
  be able to follow the code?_"
  
  * Simplify
  * Peer-review
  * Document
  
  Settling for "_just_" correct code does not 
  save you or your team any time in the long run.

]

---

class: middle

## Re-use code

We can easily re-use code within and across projects as code is separate from inputs and outputs. Three levels of re-using code:

* **Copy-paste:** Copy someone else's code and modify it to fit your context

* **Functions:** A block of code that can be referenced across a project

> * **Library:** Code packages developed by others to solve difficult tasks but made easy to use in any project

Reviewer 2: s/Library/Libraries

To be consistent with above

Author: ```suggestion
* **Libraries:** Code packages developed by others to solve difficult tasks but made easy to use in any project
```

***

> **Do not re-invent wheels**

Reviewer 2: **Do not reinvent the wheel**

https://en.wikipedia.org/wiki/Reinventing_the_wheel

Author: I know - I am referring to this saying but I am playing with words as I am not referring to a specific reinvention, but the to always be thinking this. 

If you think that is just confusing I am happy to change it to `Do not reinvent the wheel`

Reviewer 2: Up to you. I clearly missed the joke

Author: If you did then many if not most other will as well. I will keep it simple so I changed to `Do not reinvent the wheel` in de0c430a6f1ba33ed6cce3b2e2f7ff80e7c3a9cc

  ```
] 

---
class: middle

## Version control

.pull-left[

  _"Version control"_ refers to tools used to:

  * Keep track of multiple people 
  making different contributions 
  to the files of a project
  * Give access to previous versions
  of the files of a project
  
  
  Version control exists outside programming. 
  For example, `ctrl+z`, 
  restore files in OneDrive or Dropbox,
  recovered files in Windows, etc.
  
  In coding version control is much more comprehensive,
  and can track every single edit anyone in the team do.

]

.pull-left[
  .center[
    ```{r, eval=TRUE, echo=FALSE,out.width = '90%'}
      knitr::include_graphics("img/commit.png")
    ```
  ]
] 

---
class: middle

## Version control of code = Version control of all data work
  
  Code is very exact instructions on 
  how to create outputs from original data.
  
  If we handle the original data with the same level of exactness
  (backed-up or properly cited), 
  and if we version control our code, 
  then all outputs are also version controlled.
  
  Version controlling all work is perhaps
  the most valuable feature of programming.
  
  That standard for version control in code 

>   is called _Git_ or _GitHub_.

Reviewer 2: I wouldn't call git a standard, and definitely not github. Version control is much bigger than git. There's cvs and svn and others besides git. Git is just the most popular version control software and github is one of the many git repository hosting services alongside GitLab, Bitbucket, GitBucket etc.

Author: In 3c90caab91c7d1e811a265979d970722c4ba2829 I am softening the language to

>   The most common standard for version control in code is called Git and most Git projects are hosted on GitHub.

Sounds good?
